### PR TITLE
Task/11355 & task/11592: Add PointerMapper to the scheduler & remove template parameter from the PointerMapper

### PR DESCRIPTION
**Pull request by wyaneva**
This merge request relates to tasks 11355 and 11592.

We need to add a PointerMapper to the scheduler. This requires a forward declaration in the runtime. To do this, we need to remove the template argument to the PointerMapper.

* Removed the template argument to the PointerMapper class.
* Added it to methods `SYCLMalloc()` and `get_buffer()` instead.
* The buffers saved in the PointerMapper are saved as `buffer_mem` and statically casted to templated buffers in `get_buffer()`.
* Added a test for the new PointerMapper in the scheduler.

### CODE REVIEW

File: utils/vptr/include/virtual_ptr.hpp [hunk 1]
```diff
@@ -228,10 +230,15 @@ class PointerMapper {
   }
 
   /* get_buffer.
-   * Returns a buffer from the map using the pointer address
+   * Returns a pointer to a buffer from the map using the pointer address
    */
```
Comment (wyaneva) [inline utils/vptr/include/virtual_ptr.hpp]: TODO(Vanya): Remove "a pointer to" from the description.

File: utils/vptr/include/virtual_ptr.hpp [hunk 2]
```diff
@@ -44,14 +44,15 @@
...
 
 /**
  * PointerMapper
  *  Associates fake pointers with buffers.
  *
  */
-template <typename buffer_allocator =
-              cl::sycl::default_allocator<buffer_data_type> >
+// template <typename buffer_allocator =
+//              cl::sycl::default_allocator<buffer_data_type> >
```
Comment (Ruyk) [inline utils/vptr/include/virtual_ptr.hpp]: Can you remove the commented code?

File: utils/vptr/include/virtual_ptr.hpp [hunk 3]
```diff
@@ -228,10 +230,15 @@ class PointerMapper {
...
+   * Returns a pointer to a buffer from the map using the pointer address
    */
-  buffer_t get_buffer(const virtual_pointer_t ptr) {
-    return get_node(ptr)->second._b;
+  template <typename buffer_allocator = default_buffer_allocator>
+  cl::sycl::buffer<buffer_data_type, 1, buffer_allocator> get_buffer(
+      const virtual_pointer_t ptr) {
+    using buffer_t = cl::sycl::buffer<buffer_data_type, 1, buffer_allocator>;
+    auto buf = &get_node(ptr)->second._b;
+    buffer_t *b = static_cast<buffer_t *>(buf);
```
Comment (Ruyk) [inline utils/vptr/include/virtual_ptr.hpp]: Can you not simply do this?

```
buffer_t retBuffer(reinterpret_cast<buffer_t>(get_node(ptr)->second._b))
```

Comment (Ruyk) [reply utils/vptr/include/virtual_ptr.hpp]: Also, I don't think you need to pass the allocator type to the get_buffer if you use the typed allocator from the allocator headers of computecpp.

**Review by Ruyk — CHANGES_REQUESTED** [review]
Some comments in the casting of the buffer

File: utils/vptr/include/virtual_ptr.hpp (hunk 3 above)
Comment (wyaneva) [reply utils/vptr/include/virtual_ptr.hpp]: I have changed the code to use the copy constructor.

I have also removed the template argument for the allocator type.
However, to use the `typed_allocator`, I need to pass it an allocator type, so I still can't use it without one:
`template <typename T, typename AllocatorT>
class typed_allocator : public base_allocator {}`

Maybe we can use the `base_allocator` instead?

File: utils/vptr/include/virtual_ptr.hpp (hunk 2 above)
Comment (wyaneva) [reply utils/vptr/include/virtual_ptr.hpp]: Done: 17ea9e718e661f57e1773746954ceaf3509a1155

**Review by Ruyk — APPROVED** [review]

**Comment by Ruyk** [comment]
Thanks for the changes!

**Pull request merged**